File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int THRESHOLD = 10;
        HashMap<String, Integer> hm = new HashMap<>();
        for(int i = 0; i < THRESHOLD; i++){
            if(!hm.containsKey(discount[i])) hm.put(discount[i], 1);
            else hm.put(discount[i], hm.get(discount[i]) + 1);
        }
        
        HashMap<String, Integer> target = new HashMap<>();
        for(int i = 0; i < want.length; i++){
            target.put(want[i], number[i]);
        }
        
        int joinDay = 1;
        int answer = 0;
        
        // 첫날 검사
        if(check(hm, target)) answer++;
        
        while(joinDay + THRESHOLD <= discount.length){            
            hm.put(discount[joinDay - 1], hm.get(discount[joinDay - 1]) - 1);            
            if(!hm.containsKey(discount[THRESHOLD + joinDay - 1])) hm.put(discount[THRESHOLD + joinDay - 1], 1);
            else hm.put(discount[THRESHOLD + joinDay - 1], hm.get(discount[THRESHOLD + joinDay - 1]) + 1);
            
            joinDay++;
            
            if(check(hm, target)) answer++;                        
        }
        
        return answer;
    }
    
    public boolean check(HashMap<String, Integer> hm, HashMap<String, Integer> target){        
        for(String key: target.keySet()){
            if(!hm.containsKey(key) || hm.get(key) != target.get(key)) return false;
        }
        
        return true;
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
def solution(want, number, discount):
    N = len(discount)
    answer = 0
    total_amount = 0
    buy_list = {}
    
    for good, amount in zip(want, number):        
        buy_list[good] = amount
        total_amount += amount
    
    if total_amount > len(discount):
        return 0
        
    for i in range(total_amount):        
        if discount[i] in buy_list:
            buy_list[discount[i]] -= 1
    
    if is_all_bought(buy_list):
        answer += 1
    
    for i in range(1, N - total_amount + 1):
        past_good = discount[i-1]
        if past_good in buy_list:
            buy_list[past_good] += 1
            
        cur_good = discount[total_amount + i - 1]        
        if cur_good in buy_list:
            buy_list[cur_good] -= 1
            
        # print(past_good, cur_good, buy_list)
        if is_all_bought(buy_list):
            answer += 1        
    
    return answer


def is_all_bought(buy_list: dict()):
    for key in buy_list.keys():
        if buy_list[key] > 0:
            return False
    
    return True